### PR TITLE
Pytorch LR schedulers #16

### PR DESCRIPTION
**Pull request by juan43ramirez**
Closes #16 

## Changes

Adds support for Pytorch learning rate schedulers. 

Schedulers on the primal optimizer's learning rate can be handled as in standard Pytorch. I added an example on the docs for cooper.optim.

As Cooper aims to handle everything dual-variable related internally, I adopted a similar approach to the dual_scheduler as with the dual_optimizer: 
- A partial instantiation of the dual_scheduler must be provided to the __init__ of ConstrainedOptimizer. The dual_scheduler is fully instantiated internally once its respective dual_optimizer is available.
- The dual_scheduler.step() call is performed inside the constrained_optimizer.step() method, after doing a dual_optimizer.step().

Changes to the code are:
1. Refactor cooper.optim.partial to cooper.optim.partial_optimizer.
2. Add a dual_scheduler parameter to the init of ConstrainedOptimizer.
3. Create a cooper.optim.partial_scheduler (analogous to cooper.optim.partial_optimizer) for handling the dual_scheduler instantiation. 
4. Fully instantiate the dual_scheduler in constrained_optimizer.step(), after instantiating the dual_optimizer.
5. In constrained_optimizer.step() perform a dual_scheduler.step() after a dual_parameters.step(). 

Added documentation to cooper.optim for partial_scheduler and to ConstrainedOptimizer.

## Testing

1. Sanity check that a dual_scheduler is not provided to a ConstrainedOptimizer when the corresponding ConstrainedMinimizationProblem is unconstrained or no dual_optimizer was provided. 
2. test_scheduler.py ensuring that **one** call to dual_scheduler.step() happens per optimization step. This is for vanilla optimizers and extrapolation capable optimizers. I considered two torch.optim.lr_scheduler schedulers for the test. 

Note: no test was written to verify that the parameter updates match the expected updates (given that there are learning rate changes).

### CODE REVIEW

File: cooper/constrained_optimizer.py [hunk 1]
```diff
@@ -198,6 +232,12 @@ def step(
             if self.cmp.is_constrained:
                 self.dual_step()
 
+                if self.dual_scheduler is not None:
+                    # Do a step on the dual scheduler after the actual step on
+                    # the dual parameters. Extrapolation steps are not followed
```
Comment (gallego-posada) [inline cooper/constrained_optimizer.py:237]: This means that the "intermediate" updates that take place *inside* the extrapolation process do _not_ perform a call to the scheduler's step method. Correct?

**Review by gallego-posada — APPROVED** [review]
